File: lib/extract.js
// Extract text + structural hints from PDF / DOCX / PPTX / TXT / XLSX buffers.
// Returns { text, headings: [{line, level}], slides?: [string] }
import { createRequire } from 'node:module';
import mammoth from 'mammoth';
import JSZip from 'jszip';
import * as XLSX from 'xlsx';

// pdf-parse ships an unfortunate top-level test that runs on import in ESM —
// require it lazily to avoid the "ENOENT ./test/data" startup error.
const require = createRequire(import.meta.url);

export async function extractText(buffer, filetype) {
  const t = (filetype || '').toLowerCase();
  if (t.includes('pdf')) return extractPdf(buffer);
  if (t.includes('word') || t.includes('docx')) return extractDocx(buffer);
  if (t.includes('presentation') || t.includes('pptx')) return extractPptx(buffer);
  if (t.includes('spreadsheet') || t.includes('xlsx') || t.includes('xls') || t.includes('excel')) return extractExcel(buffer);
  return { text: buffer.toString('utf-8'), headings: [] };
}

async function extractPdf(buffer) {
  const pdfParse = require('pdf-parse');
  const out = await pdfParse(buffer);
  const text = (out.text || '').replace(/\r\n/g, '\n');
  const headings = detectHeadingsFromText(text);
  return { text, headings };
}

async function extractDocx(buffer) {
  // Use raw text extraction — convertToHtml drops table content silently.
  const out = await mammoth.extractRawText({ buffer });
  const text = (out.value || '').replace(/\r\n/g, '\n').replace(/\n{3,}/g, '\n\n');
  const headings = detectHeadingsFromText(text);
  return { text, headings };
}

async function extractPptx(buffer) {
  const zip = await JSZip.loadAsync(buffer);
  const slideNames = Object.keys(zip.files)
    .filter((n) => /^ppt\/slides\/slide\d+\.xml$/.test(n))
    .sort((a, b) => {
      const na = Number(a.match(/slide(\d+)\.xml$/)[1]);
      const nb = Number(b.match(/slide(\d+)\.xml$/)[1]);
      return na - nb;
    });
  const slides = [];
  for (const name of slideNames) {
    const xml = await zip.file(name).async('string');
    const matches = xml.match(/<a:t[^>]*>([^<]*)<\/a:t>/g) || [];
    const text = matches
      .map((tok) => tok.replace(/<[^>]+>/g, ''))
      .join(' ')
      .replace(/\s+/g, ' ')
      .trim();
    slides.push(text);
  }
  return { text: slides.join('\n\n'), headings: [], slides };
}

async function extractExcel(buffer) {
  const workbook = XLSX.read(buffer, { type: 'buffer' });
  const parts = [];
  for (const sheetName of workbook.SheetNames) {
    const sheet = workbook.Sheets[sheetName];
    const csv = XLSX.utils.sheet_to_csv(sheet, { blankrows: false });
    if (csv.trim()) parts.push(`Sheet: ${sheetName}\n${csv}`);
  }
  const text = parts.join('\n\n').replace(/\r\n/g, '\n');
  const headings = workbook.SheetNames.map((name, i) => ({ line: i, level: 2, text: `Sheet: ${name}` }));
  return { text, headings };
}

function stripTags(s) {
  return s
    .replace(/<[^>]+>/g, ' ')
    .replace(/&nbsp;/g, ' ')
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/\s+/g, ' ');
}

// Fallback heading detection from raw text:
//   - Markdown style:           #, ##, ### at line start
//   - Numbered sections:        "1. Title", "2.1 Subtitle"
//   - Short ALL-CAPS lines:     "OVERVIEW", "ACTION ITEMS"
//   - Title Case short lines:   "Action Items", "Decisions Made"
function detectHeadingsFromText(text) {
  const out = [];
  const lines = text.split('\n');
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();
    if (!line) continue;

    // Markdown
    const md = line.match(/^(#{1,4})\s+(.+)/);
    if (md) {
      out.push({ line: i, level: md[1].length, text: md[2].trim() });
      continue;
    }

    // Numbered: "1. Foo", "1.2 Foo", "1) Foo"
    const numbered = line.match(/^(\d+(?:\.\d+)*)[\.\)]\s+(\S.{1,80})$/);
    if (numbered && line.split(/\s+/).length <= 12 && !line.endsWith('.')) {
      out.push({ line: i, level: numbered[1].includes('.') ? 3 : 2, text: line });
      continue;
    }

    const wordCount = line.split(/\s+/).length;
    const letters = line.replace(/[^A-Za-z]/g, '');

    // ALL-CAPS short
    if (
      wordCount <= 8 &&
      letters.length >= 4 &&
      letters === letters.toUpperCase()
    ) {
      out.push({ line: i, level: 2, text: line });
      continue;
    }

    // Title Case short — every word starts uppercase, no trailing punctuation,
    // followed by a blank line (so it looks like a section header, not body text).
    if (
      wordCount >= 1 && wordCount <= 6 &&
      letters.length >= 4 &&
      !/[.!?,;:]$/.test(line) &&
      /^[A-Z]/.test(line) &&
      line.split(/\s+/).every((w) => /^[A-Z0-9]/.test(w) || w.length <= 2) &&
      (lines[i + 1] === undefined || lines[i + 1].trim() === '')
    ) {
      out.push({ line: i, level: 2, text: line });
    }
  }
  return out;
}

File: fastapi_backend/lib/extract.py
"""Text extraction from uploaded files — mirrors lib/extract.js."""
import re
import zipfile
import io
from xml.etree import ElementTree as ET


def _detect_headings(lines: list[str]) -> list[dict]:
    headings = []
    for i, line in enumerate(lines):
        stripped = line.strip()
        if not stripped:
            continue
        # Markdown headings
        m = re.match(r'^(#{1,6})\s+(.+)', stripped)
        if m:
            headings.append({"line": i, "level": len(m.group(1)), "text": m.group(2)})
            continue
        # Numbered sections 1. / 1.1 / 2.3.4
        if re.match(r'^\d+(\.\d+)*\.?\s+\S', stripped):
            headings.append({"line": i, "level": 2, "text": stripped})
            continue
        # ALL CAPS lines
        if len(stripped) >= 4 and stripped == stripped.upper() and not stripped.endswith(('.', '?', '!')):
            headings.append({"line": i, "level": 2, "text": stripped})
            continue
        # Title-case followed by blank line
        if (
            i + 1 < len(lines)
            and not lines[i + 1].strip()
            and stripped == stripped.title()
            and len(stripped.split()) <= 8
        ):
            headings.append({"line": i, "level": 3, "text": stripped})
    return headings


def extract_text(file_bytes: bytes, filetype: str) -> dict:
    ft = filetype.lower()

    if "pdf" in ft:
        return _extract_pdf(file_bytes)
    if "docx" in ft or "wordprocessingml" in ft or "openxmlformats-officedocument.wordprocessingml" in ft:
        return _extract_docx(file_bytes)
    if "pptx" in ft or "presentationml" in ft:
        return _extract_pptx(file_bytes)
    if "xlsx" in ft or "spreadsheetml" in ft or "excel" in ft:
        return _extract_xlsx(file_bytes)
    # Plain text fallback
    return _extract_text(file_bytes)


def _extract_pdf(data: bytes) -> dict:
    try:
        import fitz  # PyMuPDF
        doc = fitz.open(stream=data, filetype="pdf")
        pages = [page.get_text() for page in doc]
        text = "\n\n".join(pages)
        lines = text.splitlines()
        return {"text": text, "headings": _detect_headings(lines)}
    except Exception as e:
        return {"text": "", "headings": [], "error": str(e)}


def _extract_docx(data: bytes) -> dict:
    try:
        from docx import Document
        doc = Document(io.BytesIO(data))
        text = "\n".join(p.text for p in doc.paragraphs)
        lines = text.splitlines()
        return {"text": text, "headings": _detect_headings(lines)}
    except Exception as e:
        # Try raw XML extraction as fallback for malformed docx
        try:
            with zipfile.ZipFile(io.BytesIO(data)) as z:
                if "word/document.xml" in z.namelist():
                    xml = z.read("word/document.xml")
                    root = ET.fromstring(xml)
                    texts = [t.text or "" for t in root.iter("{http://schemas.openxmlformats.org/wordprocessingml/2006/main}t")]
                    text = " ".join(texts).strip()
                    return {"text": text, "headings": _detect_headings(text.splitlines())}
        except Exception:
            pass
        return {"text": "", "headings": [], "error": str(e)}


def _extract_pptx(data: bytes) -> dict:
    slides = []
    try:
        from pptx import Presentation
        prs = Presentation(io.BytesIO(data))
        for slide in prs.slides:
            parts = []
            for shape in slide.shapes:
                if shape.has_text_frame:
                    parts.append(shape.text_frame.text)
            slides.append(" ".join(parts).strip())
    except Exception:
        # Fallback: raw XML extraction
        try:
            with zipfile.ZipFile(io.BytesIO(data)) as z:
                names = sorted(n for n in z.namelist() if re.match(r'ppt/slides/slide\d+\.xml', n))
                for name in names:
                    xml = z.read(name)
                    root = ET.fromstring(xml)
                    ns = {'a': 'http://schemas.openxmlformats.org/drawingml/2006/main'}
                    texts = [t.text or '' for t in root.iter('{http://schemas.openxmlformats.org/drawingml/2006/main}t')]
                    slides.append(" ".join(texts).strip())
        except Exception:
            pass

    text = "\n\n".join(s for s in slides if s)
    return {"text": text, "headings": [], "slides": slides}


def _extract_xlsx(data: bytes) -> dict:
    try:
        import openpyxl
        wb = openpyxl.load_workbook(io.BytesIO(data), data_only=True)
        sheets = []
        for sheet in wb.worksheets:
            rows = []
            for row in sheet.iter_rows(values_only=True):
                cells = [str(c) if c is not None else "" for c in row]
                if any(cells):
                    rows.append(",".join(cells))
            if rows:
                sheets.append(f"Sheet: {sheet.title}\n" + "\n".join(rows))
        text = "\n\n".join(sheets)
        return {"text": text, "headings": []}
    except Exception as e:
        return {"text": "", "headings": [], "error": str(e)}


def _extract_text(data: bytes) -> dict:
    try:
        text = data.decode("utf-8", errors="replace")
    except Exception:
        text = ""
    lines = text.splitlines()
    return {"text": text, "headings": _detect_headings(lines)}
